src/github.com/getlantern/fronted/direct.go

```diff
@@ -117,6 +117,24 @@ func (d *direct) NewDirectHttpClient() *http.Client {
 	}
 }
 
+// Do continually retries a given request until it succeeds because some fronting providers
+// will return a 403 for some domains.
+func (d *direct) Do(req *http.Request) (*http.Response, error) {
+	for i := 0; i < 6; i++ {
+		client := d.NewDirectHttpClient()
+		if resp, err := client.Do(req); err != nil {
+			log.Errorf("Could not complete request %v", err)
+			continue
+		} else if resp.StatusCode > 199 && resp.StatusCode < 400 {
+			return resp, err
+		} else {
+			_ = resp.Body.Close()
+			continue
+		}
```

Comment:
Yeah that sounds good man -- I'll make that change and will then put it
into QA. I am tempted to add a test for that new Do method - might do that
in the morning. Thanks for reviewing!
On Oct 6, 2015 9:09 PM, "atavism" <notifications@github.com> wrote:

> In src/github.com/getlantern/fronted/direct.go
> <https://github.com/getlantern/lantern/pull/3239#discussion_r41349787>:
>
> > @@ -117,6 +117,24 @@ func (d *direct) NewDirectHttpClient() *http.Client {
> >  	}
> >  }
> >
> > +// Do continually retries a given request until it succeeds because some fronting providers
> > +// will return a 403 for some domains.
> > +func (d *direct) Do(req *http.Request) (*http.Response, error) {
> > +	for i := 0; i < 6; i++ {
> > +		client := d.NewDirectHttpClient()
> > +		if resp, err := client.Do(req); err != nil {
> > +			log.Errorf("Could not complete request %v", err)
> > +			continue
> > +		} else if resp.StatusCode > 199 && resp.StatusCode < 400 {
> > +			return resp, err
> > +		} else {
> > +			_ = resp.Body.Close()
> > +			continue
> > +		}
>
> Cool, do we want to just remove those continue statements then? Everything
> else looks good to me so we can go ahead and merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/pull/3239/files#r41349787>.
>